src/osp/scientific/kepler.cpp

```diff
@@ -0,0 +1,436 @@
+        double E = meanAnomaly;
+
+        // use newton's method
+        // todo: implement something which converges more quickly
+        do
+        {
+            double f_E = E - eccentricity * std::sin(E) - meanAnomaly;
+            double df_E = 1.0 - eccentricity * std::cos(E);
+
+            delta = f_E / df_E;
+            E = E - delta;
+
+            E = wrap_angle(E); // Sometimes this sometimes fails to converge without this line
+            --steps;
+        } while (std::abs(delta) > tolerance && steps > 1);
+        return E;
+    }
+
+    double solve_kepler_hyperbolic(double meanAnomaly, double eccentricity, int steps)
+    {
+        const double tolerance = 1.0E-10;
+
+        double delta;
+        double E = std::log(std::abs(meanAnomaly));
+
+        // use newton's method
+        do
+        {
+            double f_E = eccentricity * std::sinh(E) - E - meanAnomaly;
+            double df_E = eccentricity * std::cosh(E) - 1.0;
+
+            delta = f_E / df_E;
+            E = E - delta;
+            --steps;
+        } while (std::abs(delta) > tolerance && steps > 1);
+        return E;
+    }
+
+
+
+
+    double eccentric_anomaly_to_true_anomaly(const double &ecc, const double &E)
+    {
+        // This formula supposedly has better numerical stability than the one below
+        double beta = ecc / (1 + std::sqrt(1 - ecc * ecc));
+        return E + 2.0 * std::atan2(beta * std::sin(E), 1 - beta * std::cos(E));
+
+        // return 2.0 * std::atan2(std::sqrt(1 + ecc) * std::sin(E / 2.0), std::sqrt(1 - ecc) * std::cos(E / 2.0));
+    }
+
+    double hyperbolic_eccentric_anomaly_to_true_anomaly(const double &ecc, const double &H)
+    {
+        return 2 * atan(sqrt((ecc + 1) / (ecc - 1)) * tanh(H / 2));
+    }
+
+    double true_anomaly_to_eccentric_anomaly(const double &ecc, const double &v)
+    {
+        return 2.0 * std::atan(std::sqrt((1 - ecc) / (1 + ecc)) * std::tan(v / 2.0));
+    }
+
+    double true_anomaly_to_hyperbolic_eccentric_anomaly(const double &ecc, const double &v)
+    {
+        return 2.0 * std::atanh(std::clamp(std::sqrt((ecc - 1) / (ecc + 1)) * std::tan(v / 2.0), -1.0, 1.0));
+    }
+
+    double mean_motion(double mu, double a)
+    {
+        return sqrt(mu / (a * a * a));
+    }
+
+    double get_orbiting_radius(double ecc, double semiMajorAxis, double trueAnomaly)
+    {
+        return semiMajorAxis * (1 - ecc * ecc) / (1 + ecc * std::cos(trueAnomaly));
+    }
+
+
+
+
+    double KeplerOrbit::get_true_anomaly_from_eccentric(double eccentric) const
+    {
+        if (is_elliptic())
+        {
+            return eccentric_anomaly_to_true_anomaly(m_eccentricity, eccentric);
+        }
+        else if (is_hyperbolic())
+        {
+            return hyperbolic_eccentric_anomaly_to_true_anomaly(m_eccentricity, eccentric);
+        }
+    }
+
+    double KeplerOrbit::get_eccentric_anomaly(double time) const
+    {
+        if (is_elliptic())
+        {
+#if 1 // Use standard kepler solver
```

Comment:
whatever is going to be here in the final version, `if constexpr` can likely do the job here